# Day 1 Reading Journal

Hi everyone.  These are the exercises for the first set of readings from ThinkPython.  While this journal consists of several required exercises, it is meant to encourage active reading more generally.  You should use this journal to take detailed notes, catalog questions, and explore the content from ThinkPython deeply.

For more information on using the Jupyter notebook, check out the documentation [here](http://nbviewer.jupyter.org/github/jupyter/notebook/blob/master/docs/source/examples/Notebook/Examples%20and%20Tutorials%20Index.ipynb).  Additionally, in order to write text in this document, you will be making use of markdown cells. A handy cheatsheet for markdown is available [here](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet).

## Chapter 1

Note: the exercise numbers below match up with the reading for ease of cross referencing.

### Python Programming Language
#### High-Level Language:
Requires extra processing time

Easier to write

Interpreters and Compilers convert high-level into low-level language

"Source Code" and "Executable" are names for converted source code

#### Low-level Language:
"Machine Language" because can be run by a machine without processing

Not easily interchangeable between computers.

#### Interpreter:
Interactive Mode uses prompt ">>>"

Script mode stores code until you run it

### What is a program?
#### Programs:
Sequence of instructions (e.g. input, output, math, conditional execution, repetition)

### What is Debugging?
#### Debugging:
Errors are called bugs and three kinds can occur:

-Syntax

    Structure of the program
    
-Runtime

    Error occurs in the middle of running program
    
-Semantic

    Program runs, but you wrote it wrong
    
Modify program and try again to debug

### Formal and Natural Languages
Natural languages are evolved naturally and not design

-Have lots of ambiguity that depend on contextual clues

-Redundant to make up for ambiguity

-Literalness

Formal languages are designed for specific purposes e.g. math

-Concise and unambiguous

"Programming languages are formal languages that have been designed to express computations."
#### Ambiguity
From most to least ambiguous: Poetry, Prose, Programs
#### Syntax
Syntax strict in formal languages

Tokens: Actual elements of the language

Structure: Arrangement of tokens

Parsing: Interpreting structure of a "sentence"

### Exercise 3

Type `help()` to start the online help utility. Or you can type help('print') to get information about the print statement.  You should type `q` and then hit `enter` in the text box to exit the help utility. 

Note: this exercise is pretty simple (and there's not much to put in the box)!  We just want to make sure that you have tried out this super-handy feature of Python!

In [1]:
help()


Welcome to Python 2.7!  This is the online help utility.

If this is your first time using Python, you should definitely check out
the tutorial on the Internet at http://docs.python.org/2.7/tutorial/.

Enter the name of any module, keyword, or topic to get help on writing
Python programs and using Python modules.  To quit this help utility and
return to the interpreter, just type "quit".

To get a list of available modules, keywords, or topics, type "modules",
"keywords", or "topics".  Each module also comes with a one-line summary
of what it does; to list the modules whose summaries contain a given word
such as "spam", type "modules spam".

help> q

You are now leaving help and returning to the Python interpreter.
If you want to ask for help on a particular object directly from the
interpreter, you can type "help(object)".  Executing "help('string')"
has the same effect as typing a particular string at the help> prompt.


### Exercise 4  

Start the Python interpreter and use it as a calculator. Python’s syntax for math operations is almost the same as standard mathematical notation. For example, the symbols +, - and / denote addition, subtraction and division, as you would expect. The symbol for multiplication is *.

If you run a 10 kilometer race in 43 minutes 30 seconds, what is your average time per mile? What is your average speed in miles per hour? (Hint: there are 1.61 kilometers in a mile). 

In [21]:
(43.0 * 60.0 + 30.0)/(10/1.61)

420.21000000000004

420.21 seconds/mile

In [22]:
(10.0/1.61)/((43.0 * 60.0 + 30.0)/(60.0 * 60.0))

8.567144998929106

8.567 miles/hr

## Chapter 2

### Values and Types
Value is basic info a program uses (str, int, float)

No commas inside a number
### Variables
Name that refers to a value

Assignment statement creates variables and gives them a value

State Diagram-Write name with arrow pointing at variable's value
### Variable Names and keywords
Should begin with letter and lowercase (unpreferably uppercase)

Underscore for good style

Illegal variable name gives syntax error

Keyword is a word that describes structure of program and can't be used for variable names

(and as assert break class continue def del elif else except exec finally for from global if import in is lambda not or pass print raise return try while with yield)

### Operators and Operands
Operators and symbols which represent computations

Operands are values operator is applied to

Floor division rounds values down unless floats are used

### Expressions and statements
Expressions are combinations of values, variables, and operators--can even be just one of them

Statement is unit of code that Python can run
### Interactice mode and script mode
Script will not print out values unless you tell Python to do so

### String Operations
Concatenation of strings with +
Repetition of strings with *<#>

### Comments
'#' gives comments

### Debugging
Case sensitive values




### Exercise 2  

Assume that we execute the following assignment statements:

width = 17
height = 12.0
delimiter = '.'

For each of the following expressions, write the value of the expression and the type (of the value of the expression).

1. `width/2`
2. `width/2.0`
3. `height/3`
4. `1 + 2 * 5`
5. `delimiter * 5`

Please use the following markdown cell to provide your answer.

1. 8 <int>
2. 8.5 <float>
3. 4.0 <float>
4. 11 <int>
5. ..... <str>

### Exercise 3  

Practice using the Python interpreter as a calculator.  For each of these provide Python code that computes each answer.

    


The volume of a sphere with radius r is 4/3 $\pi r^3$. What is the volume of a sphere with radius 5? Hint: 392.7 is wrong!

In [15]:
(4.0/3.0 * pi * 5**3)

523.3333333333334

Suppose the cover price of a book is \$24.95, but bookstores get a 40% discount. Shipping costs \$3 for the first copy and 75 cents for each additional copy. What is the total wholesale cost for 60 copies?

In [16]:
60 * (.75 + 24.95 * .6) + 3

946.1999999999999

If I leave my house at 6:52 am and run 1 mile at an easy pace (8:15 per mile), then 3 miles at tempo (7:12 per mile) and 1 mile at easy pace again, what time do I get home for breakfast? 

In [20]:
(2*(15.0/60 + 8) + 3*(12.0/60 + 7) + 6*60 + 52)/60

7.501666666666667

Around 7:30

## Chapter 3

### Function Calls
Functions are statements which perform a specific computation

Value in parantheses is argument

### Type Conversion Function
float() converts ints and strings to floats

string() converts argument to string

int() converts floats to ints by chopping off decimal component, not rounding!!!!!

### Math functions
module is a file that contains a collection of related functions

"import math" needed first and creates module object "math"

math.*function* calls a math function -- called dot notation

functions include:

log10, log *(base e)*, sin, cos, pi, sqrt

### Adding New Functions
"def" defines a keyword and requires parantheses at the end

First line of function definition is called "header" and the rest is a "body"

Header ends with colon and body is indented

Print statements generally use double quote

Defining a function creates a variable with the name of type function. Value of the function is a function object.

### Parameters and Arguments
Parameters are the values that take the place of an argument

Argument inside function call is applied before it runs the function

### Stack Diagrams
Stack diagram show value of each variable and also shows the function each variable is a part of.

-Function is represented by a frame

-Name of function next to frame

-Parameters and Variables inside frame

-main is the topmost frame and any variable outside of a function is in main

### Importing with from
Using "from" in front of an import allows you to access specific objects from that module directly. e.g. pi

Adding an * at the end import every object from that module directly.

### Debugging
Use spaces instead of tabs if no tabs are available in text editor

In [35]:
b()

def a():
    print "fdffd"
    print "dffsf"
    
def b():
    a()
    a()

fdffd
dffsf
fdffd
dffsf


In [34]:
def b():
    a()
    a()
    
def a():
    print "fdffd"
    print "dffsf"
    
b()

fdffd
dffsf
fdffd
dffsf


### Exercise 3

Python provides a built-in function called len that returns the length of a string, so the value of len('allen') is 5.
Write a function named right_justify that takes a string named s as a parameter and prints the string with enough leading spaces so that the last letter of the string is in column 70 of the display.

```
>>> right_justify('allen')
    
                                                                 allen```
                                                                 
def right_justify(s):
    print(' '*(70-len(s)) + s)

In [45]:
def right_justify(s):
    print(' '*(70-len(s)) + s)
    
right_justify(s);

                                                                     s


### Exercise 5

This exercise can be done using only the statements and other features we have learned so far.

(a) Write a function that draws a grid like the following:
```
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
```
**Hint:** to print more than one value on a line, you can print a comma-separated sequence:
print '+', '-'
If the sequence ends with a comma, Python leaves the line unfinished, so the value printed next appears on the same line.
print '+', 
print '-'
The output of these statements is '+ -'.
A print statement all by itself ends the current line and goes to the next line.

In [80]:
def draw_window():
    draw_plus();
    draw_lines();
    draw_lines();
    draw_lines();
    draw_lines();
    draw_plus();
    draw_lines();
    draw_lines();
    draw_lines();
    draw_lines();
    draw_plus();
    
def draw_lines():
    print('|' + ' '*9 + '|' + ' '*9 + '|')
    
def draw_plus():
    print('+ ' + '- '*4 + '+ ' + '- '*4 + '+')
    
draw_window();

+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +


(b) Write a function that draws a similar grid with four rows and four columns.

In [62]:
def draw_window():
    draw_plus();
    draw_lines();
    draw_lines();
    draw_plus();
    draw_lines();
    draw_lines();
    draw_plus();
    draw_lines();
    draw_lines()
    draw_plus();
    draw_lines();
    draw_lines();
    draw_plus();
    
def draw_lines():
    print('|' + ' '*5 + '|' + ' '*5 + '|'+ ' '*5 + '|')
    
def draw_plus():
    print('+ ' + '- '*2 + '+ ' + '- '*2 + '+ ' + '- '*2 + '+')
    
draw_window();

+ - - + - - + - - +
|     |     |     |
|     |     |     |
+ - - + - - + - - +
|     |     |     |
|     |     |     |
+ - - + - - + - - +
|     |     |     |
|     |     |     |
+ - - + - - + - - +
|     |     |     |
|     |     |     |
+ - - + - - + - - +


## Chapter 5.1 - 5.7

### Modulus Operator
Modulus Operator gives remainders: 7%3 = 1

### Logical Operators
and, not, or

### Conditional Execution
If statement implementation:

if x > 0:

Boolean expression after if is the condition. Compound statements have the same structure as a function (i.e. header followed by indented body)

### Alternative Execution
if x > 0:

    fdjffe
    
else:

    fdfee
    
Alternatives are called branches

### Chained Conditionals
if x> 0:

    fdfef
    
elif x < 0:
    
    dfef

### Exercise 3  

Fermat’s Last Theorem says that there are no positive integers a, b, and c such that $a^n + b^n = c^n$ for any values of n greater than 2.

(a) Write a function named `check_fermat` that takes four parameters—a, b, c and n—and that checks to see if Fermat’s theorem holds. If n is greater than 2 and it turns out to be true that
$a^n + b^n = c^n$ the program should print, "Holy smokes, Fermat was wrong!" Otherwise the program should print, "No, that doesn’t work."

In [100]:
def check_fermat(a,b,c,n):
    test = (a**n + b**n == c**n);
    if test and n > 2:
        print("Holy smokes, Fermat was wrong!")
    else:
        print("No, that doesn’t work.")
        
check_fermat(1,1,2,3);

No, that doesn’t work.


(b) Write a function that prompts the user to input values for a, b, c and n, converts them to integers, and uses check_fermat to check whether they violate Fermat’s theorem.

In [101]:
def check_fermat(a,b,c,n):
    test = (a**n + b**n == c**n);
    if test and n > 2:
        print("Holy smokes, Fermat was wrong!")
    else:
        print("No, that doesn’t work.")

def convert(a,b,c,n):
    a = int(a)
    b = int(b)
    c = int(c)
    n = int(n)
    check_fermat(a,b,c,n)
    
convert(1,2,3,4)

No, that doesn’t work.


### Exercise 4  

If you are given three sticks, you may or may not be able to arrange them in a triangle. For example, if one of the sticks is 12 inches long and the other two are one inch long, it is clear that you will not be able to get the short sticks to meet in the middle. For any three lengths, there is a simple test to see if it is possible to form a triangle:
> If any of the three lengths is greater than the sum of the other two, then you cannot form a triangle. Otherwise, you can. (If the sum of two lengths equals the third, they form what is called a “degenerate” triangle.)

(a) Write a function named `is_triangle` that takes three integers as arguments, and that prints either "Yes" or "No," depending on whether you can or cannot form a triangle from sticks with the given lengths.

In [95]:
def is_triangle(a,b,c):
    if (a + b >= c):
        if (b + c >= a):
            if (c + a >= b):
                print('Yes')
            else:
                print('No')
        else:
            print('No')
    else:
        print('No')
        
is_triangle(1,3,4)

Yes


(b) Write a function that prompts the user to input three stick lengths, converts them to integers, and uses is_triangle to check whether sticks with the given lengths can form a triangle.

In [99]:
def convert(a,b,c):
    a = int(a)
    b = int(b)
    c = int(c)
    is_triangle(a,b,c);
    
convert(2.4, 3.7, 6.1)

No


## Notes for the Instructors

Please use the space below to write comments to help us plan the next class session.  For instance, if you want to see us go over an example of a particular type of problem, you can indicate that here.

Please remember that the best way to get quick feedback from the instructors as well as your peers, is to use Piazza.  However, consider using this channel if it doesn't make sense to post your note to Piazza.